### ISSUE TITLE
Allow multithreading for "AsNoTracking()"

### ISSUE THREAD

**Issue by angelaki** [dotnet/efcore#18094]
Parallel execution of DbContext requests.

If I'm not mistaken I guess it would be no problem to allow EF to handle multiple parallel requets on the same context, as long as all request are not tracked.

```C#
var elementAs = await _service.GetAs().ToArrayAsync();
var elementBs = await _service.GetBs().ToArrayAsync();
var elementCs = await _service.GetCs().ToArrayAsync();
var elementDs = await _service.GetDs().ToArrayAsync();
```

works, but

`Task.WaitAll(_service.GetAs().ToArrayAsync(), _service.GetBs().ToArrayAsync(), _service.GetCs().ToArrayAsync(), _service.GetDs().ToArrayAsync());`

throws an InvalidOperationException. As far as I know there is no nice solution for this. When just resolving the services by an interface I do not know if they use EF or not in their implementation so I **need** to assume that all services aren't thread safe.

Is there a reason that this is not possible:

`Task.WaitAll(_service.GetAs().AsNoTracking().ToArrayAsync(), _service.GetBs().AsNoTracking().ToArrayAsync(), _service.GetCs().AsNoTracking().ToArrayAsync(), _service.GetDs().AsNoTracking().ToArrayAsync());`

**Comment by roji** [comment]
Concurrent use of the same DbContext is not possible - and not just for tracked queries. Specifically, a DbContext has an underlying DbConnection to the database, which cannot be used concurrently. There are other various components at work under the hood which don't support multithreading.

However, there's nothing wrong with instantiating several DbContexts and executing queries on them - whether tracking or non-tracking. That should get you the behavior you're looking for. If you run into any further issues don't hesitate to post back.

**Comment by angelaki** [comment]
Thank you very much for you answer. Ok, now I understand why concurrency is not easily implementable. But eventhough this is not necessarily an EF Core specific problem: is there a "best practice" for using EF Core with multiple contexts in a DI controlled service, where one instance gets injected / disposed? Just creating own instances and dispose them in the services' dispose event?

**Comment by roji** [comment]
@angelaki see #4441. The current best practice for this is simply to accept your `DbContextOptions<TContext>` via DI and create DbContexts with new, passing the options to them. This leaves the configuration management in DI as usual but allows you to instantiate DbContexts at will and manage their lifetime. There have been discussions on having a factory to contexts, but that simply didn't seem valuable enough.

**Comment by angelaki** [comment]
@roji Using it's constructor  (new()) I loose the ability of using DI for my DbContexts. Is there a way that keeps this ability? Createing them via ServiceProvider (event inside new scope) returns equal instances. Activating them via ActivatorUtilities works, but disposing them in the Services' dispose method often gets called too early (before AspNetCore serialization).

**Comment by roji** [comment]
> I loose the ability of using DI for my DbContexts. Is there a way that keeps this ability?

If I'm understand you correctly, you want to have multiple DbContexts which automatically get disposed when the scope ends. If so, and you're encountering an ordering issue with the disposal (with AspNetCore serialization), I'm not really sure what else is possible... However, maybe @smitpatel or @ajcvickers (on vacation) have some ideas.

**Comment by smitpatel** [comment]
> Using it's constructor (new()) I loose the ability of using DI for my DbContexts. Is there a way that keeps this ability?

I am not sure if I understand this. If you register your DbContext in DI then DbContextOptions are also registered in DI. Whenever you ask DI to inject DbContext it will do so. Since you need to create multiple DbContext, you ask for DbContextOptions & new-up yourself. It does not block you from inject DbContext directly. It is actually simpler pattern then creating so many scopes.

**Comment by angelaki** [comment]
@smitpatel But when to dispose them? Right now I've implemented IDisposable in my Service:

 - ApiController gets injected / calls Service
 - Service creates / call EFCore and responds with IQueryable
 - Service implements IDisposable and calls Context's dispose when Services disposes

Right now often the Context seams to get disposed **before** Controller has successfully serialized IQueryable. I often get an error "can not access disposed object". So when to dispose my manually created Contexts?

**Comment by roji** [comment]
@angelaki to be sure we understand your scenario...

* DI provides one "main" injected scoped DbContext, which you can use successfully and whose IQueryables you can return from the controller.
* You want to execute several queries in parallel (e.g. for efficiency), so you want to have more than one DbContext to service a single request.
* Your controller method ultimately returns a single IQueryable. Therefore, is there a reason you can't simply return the main (scoped) DbContext, provided by default by DI, from the controller, instead of any of the additional DbContexts? This way you're responsible for the lifetime of the additional ones - creating them with new and disposing them yourself, *within* your controller method.

Perhaps a small code sample could help illustrate your scenario.

**Comment by angelaki** [comment]
@roji Actually the scenario is a bit different:

The controller calls multiple service methods and awaits them parallel:

var result1Task = service.GetResult1Async();
var result2Task = service.GetResult2Async();
var result3Task = service.GetResult3Async();
Task.WaitAll(result1Task, result2Task, result3Task);

Neither does the service know if it get's called simultaneously nor do I want to implement / block it. That's why I think using injected Contexts is a bad pattern in general since it can misbehave if the service is used concurrent - and I do not want to restrict all services to be used just synchronously.

But yes, if I use the scoped DbContext it does not get disposed too soon.

**Comment by angelaki** [comment]
Yes, that is absolutely correct. But if I just implement the service I can not ensure that it is used only sequential and not parallel - neither do I want to say "my service may only be used sequntial!".

But my created contexts with "new()", disposed in my service's dispose-method just get disposed too soon! While the container controled context (DI injected) is not. My question is just: when to dispose a manually created context so that it just behaves like the container controlled one.

**Comment by roji** [comment]
> But if I just implement the service I can not ensure that it is used only sequential and not parallel - neither do I want to say "my service may only be used sequntial!".

Sorry, this isn't very clear to me... It's the method developer's responsibility to use the DbContext instance correctly. Usually that means in sequence (with multiple awaits). The service or DI have nothing to do with that.

> My question is just: when to dispose a manually created context so that it just behaves like the container controlled one.

I don't know enough about DI and its interaction with ASP.NET to answer this. But I'm proposing that you approach this another way: launch your queries with your new-ed up contexts, await the results, consume those results and then dispose the contexts - all in the same method, before returning. You can then take the consumed query results and join them together, or return them for JSON serialization, or whatever. No need for the contexts to stay alive.

**Comment by angelaki** [comment]
Sorry, I think this is getting more and more an AspNet DI topic and I should maybe ask just in another repo. When implementing the service I do not know if someone is going to use my service concurrent. Assuming this, I may never use the DI resolved Context if I want to return Queryables - and that's what I do in my case. So just materializing the queries on return would surely fix my issue, but I like the idea of returning queryables, where the person using my service can still restrict my resultset before the query gets executed (in my case by the JSON serializer. Materializing in the controller would be a solution but is not possible with the "regular" resolved Context that just somehow gets disposed late enough).

I hope this is not getting of topic? If it does you may just close this issue, thank you very much for your support anyway!

**Comment by roji** [comment]
No worries and even if we're a bit out of topic that's not so bad :)

> Assuming this, I may never use the DI resolved Context if I want to return Queryables

I disagree - the overwhelmingly common case is to use a single DbContext to service a request. Also, the service in question - DbContext - is (hopefully) clearly documented as non-thread-safe. Because of these two things the current pattern works well for almost everyone. Your scenario is definitely advanced/out of the ordinary, and it can also be resolved easily via extra contexts (even if they cannot be returned and must be manually disposed).

To summarize, if you're looking to return *multiple* IQueryables that will be evaluated/materialized *after* your method returns, then you're right that we'd need to dig further into DI, ASP.NET and how to have multiple contexts bound to the scope that get disposed at the right time. I don't have an answer for that.

But at the end of the day it's worth examining very carefully whether you really need query parallelism in your ASP.NET controller. This is quite atypical, and as I wrote above may ironically have adverse perf consequences (database connections and resources are scarce).

**Comment by angelaki** [comment]
I've tried a factory-like pattern now. First the created context's dispose was called in my service - what caused them getting disposed too soon. So I thought: "When I want the default behavior of a context getting resolved, I should hold them in an object getting resolved, too". Than I moved my code to a factory-like pattern implementing IDisposable but still the same:

`public interface IDbContextFactory
    { T CreateDbContext<T>() where T : Microsoft.EntityFrameworkCore.DbContext; }

    public class DbContextFactory : IDbContextFactory, IDisposable
    {
        private IServiceProvider _serviceProvider;

        public DbContextFactory(IServiceProvider paramServiceProvider)
        { _serviceProvider = paramServiceProvider; }

        private static List<Microsoft.EntityFrameworkCore.DbContext> _createdDbContexts = new List<Microsoft.EntityFrameworkCore.DbContext>();
        public T CreateDbContext<T>() where T : Microsoft.EntityFrameworkCore.DbContext
        {
            var result = ActivatorUtilities.CreateInstance<T>(_serviceProvider);
            _createdDbContexts.Add(result);
            return result;
        }

        private bool _disposed = false;
        protected virtual void Dispose(bool paramDisposing)
        {
            if (!_disposed)
            {
                //If uncommeted, System.ObjectDisposedException
                //if (paramDisposing) { foreach (var dbc in _createdDbContexts) { dbc.Dispose(); } }
                _disposed = true;
            }
        }

        public void Dispose() => Dispose(true);
    }`

And idea how to hook / listen to the same event that would get called for the "single DI controlled" context?

**Comment by smitpatel** [comment]
If you are using service like this `service.GetResult1Async();` and if your service depends on DbContext then you need to create different instances of your service in your controller. DbContext is not thread-safe. So your service which depends on DbContext is not thread-safe either. A different instance of the service which contains different instance of DbContext can run in parallel. Same instance of service cannot run in parallel, just the way you are hitting issue with DbContext in parallel right now. So in order to use Task.WhenAll in your controller, you need to new-up your service with new DbContext & use those, then you wouldn't run into access to disposed object issues.

**Comment by angelaki** [comment]
@smitpatel This solution is ok for me. But this means I may **never** use any service concurrent since I do not know if the service I consume is based on EF or not. Since I do not like this idea I'm looking for a solution for my service being able to get used parallel.

**Comment by smitpatel** [comment]
You have 2 options
- If your service depends on something which is not thread-safe then make your service also non-thread-safe.
- Or, make your service to introduce thread-safety so that you can consume non-thread-safe services inside.

You cannot have it both ways. i.e. Using your service in parallel but also having delayed execution from DbContext.

**Comment by angelaki** [comment]
But with my latest factory-like pattern I'm pretty close to what I'd like to achieve! If only the dispose wasn't called too soon or I'd hook to the right event I'd be pretty happy with it :)

**Comment by roji** [comment]
> This solution is ok for me. But this means I may never use any service concurrent since I do not know if the service I consume is based on EF or not. Since I do not like this idea I'm looking for a solution for my service being able to get used parallel.

Components are supposed to advertise or document whether they are thread-safe or not. You are not meant to assume that anything is threadsafe, nor to assume that nothing is threadsafe - some services are indeed threadsafe. Just look at the documentation of the particular component you want to use.

**Comment by angelaki** [comment]
But is there no way to somehow get the moment the regular DbContext gets disposed? Or can I get the request itself injected and subscribe to it's end event? But anyway this would mean running code from a disposed object - not that cool. Sorry for getting more and more off-topic. But being this close to a nice solution I don't like to stop digging deeper now.

**Comment by roji** [comment]
As I said, I I don't know much about that side of things (DI scope disposal timing)...

**Comment by angelaki** [comment]
Can you tell me what way the extension method "AddDbContext" adds the context to the service collection? Ok, bit lazy - sure I could look up myself ...
